File: totoggle_go/internal/serverapi/fetch_test.go
package serverapi

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/manorfm/toToggles/totoggle_go/internal/toggle"
)

func mustPath(t *testing.T, raw string) toggle.Path {
	t.Helper()
	p, err := toggle.NewPath(raw)
	require.NoError(t, err)
	return p
}

func TestFetcher_Fetch_SendsSecretKeyHeaderAndParsesToggles(t *testing.T) {
	var gotHeader string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotHeader = r.Header.Get("X-API-Key")
		assert.Equal(t, http.MethodGet, r.Method)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{
			"application": {
				"id": "app-1",
				"name": "Checkout Web",
				"toggles": [
					{"id": "t1", "path": "user", "value": "user", "enabled": true, "level": 0,
					 "parent_id": null, "app_id": "app-1", "has_activation_rule": false, "activation_rule": null},
					{"id": "t2", "path": "user.payments", "value": "payments", "enabled": true, "level": 1,
					 "parent_id": "t1", "app_id": "app-1", "has_activation_rule": true,
					 "activation_rule": {"type": "percentage", "value": "50"}}
				]
			}
		}`))
	}))
	defer srv.Close()

	fetcher := NewFetcher(nil, srv.URL, "sk_test123")
	app, err := fetcher.Fetch(context.Background())

	require.NoError(t, err)
	assert.Equal(t, "sk_test123", gotHeader)
	require.Len(t, app.Toggles, 2)

	t2, ok := app.ByPath(mustPath(t, "user.payments"))
	require.True(t, ok)
	require.NotNil(t, t2.ActivationRule)
	assert.Equal(t, "50", t2.ActivationRule.Value)
}

func TestFetcher_Fetch_UnauthorizedReturnsErrAuthentication(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
	}))
	defer srv.Close()

	fetcher := NewFetcher(nil, srv.URL, "sk_bad")
	_, err := fetcher.Fetch(context.Background())

	require.Error(t, err)
	assert.True(t, errors.Is(err, ErrAuthentication))
}

// A secret key rejection must never appear verbatim in the returned error — only the sentinel,
// so a caller who logs err.Error() can't leak the key into their own logs.
func TestFetcher_Fetch_ErrorNeverLeaksTheSecretKey(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
	}))
	defer srv.Close()

	fetcher := NewFetcher(nil, srv.URL, "sk_supersecret")
	_, err := fetcher.Fetch(context.Background())

	require.Error(t, err)
	assert.NotContains(t, err.Error(), "sk_supersecret")
}

func TestFetcher_Fetch_ServerErrorStatus(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
	}))
	defer srv.Close()

	fetcher := NewFetcher(nil, srv.URL, "sk_test")
	_, err := fetcher.Fetch(context.Background())
	require.Error(t, err)
}

func TestFetcher_Fetch_MalformedJSON(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`not json`))
	}))
	defer srv.Close()

	fetcher := NewFetcher(nil, srv.URL, "sk_test")
	_, err := fetcher.Fetch(context.Background())
	require.Error(t, err)
}

func TestFetcher_Fetch_RespectsContextCancellation(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(50 * time.Millisecond)
	}))
	defer srv.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Millisecond)
	defer cancel()

	fetcher := NewFetcher(nil, srv.URL, "sk_test")
	_, err := fetcher.Fetch(ctx)
	require.Error(t, err)
}

File: totoggle_node/src/internal/serverapi/fetch.test.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";
import { afterEach, describe, expect, it } from "vitest";
import { TotoggleAuthenticationError } from "../../errors.js";
import { Path } from "../toggle/path.js";
import { fetchToggles } from "./fetch.js";

let server: Server | undefined;

function listen(handler: (req: IncomingMessage, res: ServerResponse) => void): Promise<string> {
  server = createServer(handler);
  return new Promise((resolve) => {
    server!.listen(0, "127.0.0.1", () => {
      const address = server!.address();
      if (address === null || typeof address === "string") {
        throw new Error("unexpected server address");
      }
      resolve(`http://127.0.0.1:${address.port}`);
    });
  });
}

afterEach(() => {
  server?.close();
  server = undefined;
});

describe("fetchToggles", () => {
  it("sends the X-API-Key header and parses the toggle set", async () => {
    let receivedKey: string | undefined;
    const url = await listen((req, res) => {
      receivedKey = req.headers["x-api-key"] as string;
      res.setHeader("Content-Type", "application/json");
      res.end(
        JSON.stringify({
          application: {
            id: "app-1",
            name: "Checkout Web",
            toggles: [
              {
                id: "t1",
                path: "user",
                value: "user",
                enabled: true,
                level: 0,
                parent_id: null,
                app_id: "app-1",
                has_activation_rule: false,
                activation_rule: null,
              },
              {
                id: "t2",
                path: "user.payments",
                value: "payments",
                enabled: true,
                level: 1,
                parent_id: "t1",
                app_id: "app-1",
                has_activation_rule: true,
                activation_rule: { type: "percentage", value: "50" },
              },
            ],
          },
        }),
      );
    });

    const app = await fetchToggles(`${url}/api/toggles`, "sk_test123", 1000);

    expect(receivedKey).toBe("sk_test123");
    expect(app.toggles).toHaveLength(2);
    const t2 = app.byPath(Path.parse("user.payments"));
    expect(t2?.activationRule).toEqual({ type: "percentage", value: "50" });
  });

  it("throws TotoggleAuthenticationError on 404", async () => {
    const url = await listen((_req, res) => {
      res.writeHead(404);
      res.end();
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_bad", 1000)).rejects.toBeInstanceOf(
      TotoggleAuthenticationError,
    );
  });

  it("throws TotoggleAuthenticationError on 401", async () => {
    const url = await listen((_req, res) => {
      res.writeHead(401);
      res.end();
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_bad", 1000)).rejects.toBeInstanceOf(
      TotoggleAuthenticationError,
    );
  });

  // The rejection must never include the secret key verbatim, so a caller logging err.message
  // can't leak the credential.
  it("never leaks the secret key in the error", async () => {
    const url = await listen((_req, res) => {
      res.writeHead(401);
      res.end();
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_supersecret", 1000)).rejects.not.toThrow(
      /sk_supersecret/,
    );
  });

  it("throws on a non-2xx, non-auth status", async () => {
    const url = await listen((_req, res) => {
      res.writeHead(500);
      res.end();
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_test", 1000)).rejects.toThrow();
  });

  it("throws on malformed JSON", async () => {
    const url = await listen((_req, res) => {
      res.end("not json");
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_test", 1000)).rejects.toThrow();
  });

  it("throws when the response has no \"application\" object", async () => {
    const url = await listen((_req, res) => {
      res.setHeader("Content-Type", "application/json");
      res.end(JSON.stringify({ notWhatWeExpected: true }));
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_test", 1000)).rejects.toThrow(
      /application/,
    );
  });

  it("throws when application.toggles isn't an array", async () => {
    const url = await listen((_req, res) => {
      res.setHeader("Content-Type", "application/json");
      res.end(JSON.stringify({ application: { id: "app-1", name: "x", toggles: "not-an-array" } }));
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_test", 1000)).rejects.toThrow(/toggles/);
  });

  it("times out via AbortController when the server hangs", async () => {
    const url = await listen(() => {
      // never respond
    });

    await expect(fetchToggles(`${url}/api/toggles`, "sk_test", 20)).rejects.toThrow();
  });
});
